In [1]:
import plotly.express as px
import pandas as pd
import matplotlib.pyplot as plt
from pandasql import sqldf
import numpy as np

In [2]:
url = "https://raw.githubusercontent.com/PhiPhuongUyen/Team-5/main/marketing_campaign.csv"
df= pd.read_csv(url, sep="\t")
pysqldf = lambda q: sqldf(q, globals())
df

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
0      5524        1957  Graduation         Single  58138.0        0   
1      2174        1954  Graduation         Single  46344.0        1   
2      4141        1965  Graduation       Together  71613.0        0   
3      6182        1984  Graduation       Together  26646.0        1   
4      5324        1981         PhD        Married  58293.0        1   
...     ...         ...         ...            ...      ...      ...   
2235  10870        1967  Graduation        Married  61223.0        0   
2236   4001        1946         PhD       Together  64014.0        2   
2237   7270        1981  Graduation       Divorced  56981.0        0   
2238   8235        1956      Master       Together  69245.0        0   
2239   9405        1954         PhD        Married  52869.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
0            0  04-09-2012       58       635  ...                  7   
1            1  08-03-2014       38        11  ...                  5   
2            0  21-08-2013       26       426  ...                  4   
3            0  10-02-2014       26        11  ...                  6   
4            0  19-01-2014       94       173  ...                  5   
...        ...         ...      ...       ...  ...                ...   
2235         1  13-06-2013       46       709  ...                  5   
2236         1  10-06-2014       56       406  ...                  7   
2237         0  25-01-2014       91       908  ...                  6   
2238         1  24-01-2014        8       428  ...                  3   
2239         1  15-10-2012       40        84  ...                  7   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
0                0             0             0             0             0   
1                0             0             0             0             0   
2                0             0             0             0             0   
3                0             0             0             0             0   
4                0             0             0             0             0   
...            ...           ...           ...           ...           ...   
2235             0             0             0             0             0   
2236             0             0             0             1             0   
2237             0             1             0             0             0   
2238             0             0             0             0             0   
2239             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
0            0              3         11         1  
1            0              3         11         0  
2            0              3         11         0  
3            0              3         11         0  
4            0              3         11         0  
...        ...            ...        ...       ...  
2235         0              3         11         0  
2236         0              3         11         0  
2237         0              3         11         0  
2238         0              3         11         0  
2239         0              3         11         1  

[2240 rows x 29 columns]

In [3]:
df1=pysqldf("SELECT ID, Year_Birth, Income, Dt_Customer, MntWines, MntFruits, MntMeatProducts, MntFishProducts, MntSweetProducts, MntGoldProds\
             FROM df")

df1

ID  Year_Birth   Income Dt_Customer  MntWines  MntFruits  \
0      5524        1957  58138.0  04-09-2012       635         88   
1      2174        1954  46344.0  08-03-2014        11          1   
2      4141        1965  71613.0  21-08-2013       426         49   
3      6182        1984  26646.0  10-02-2014        11          4   
4      5324        1981  58293.0  19-01-2014       173         43   
...     ...         ...      ...         ...       ...        ...   
2235  10870        1967  61223.0  13-06-2013       709         43   
2236   4001        1946  64014.0  10-06-2014       406          0   
2237   7270        1981  56981.0  25-01-2014       908         48   
2238   8235        1956  69245.0  24-01-2014       428         30   
2239   9405        1954  52869.0  15-10-2012        84          3   

      MntMeatProducts  MntFishProducts  MntSweetProducts  MntGoldProds  
0                 546              172                88            88  
1                   6                2                 1             6  
2                 127              111                21            42  
3                  20               10                 3             5  
4                 118               46                27            15  
...               ...              ...               ...           ...  
2235              182               42               118           247  
2236               30                0                 0             8  
2237              217               32                12            24  
2238              214               80                30            61  
2239               61                2                 1            21  

[2240 rows x 10 columns]

In [4]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ID                2240 non-null   int64  
 1   Year_Birth        2240 non-null   int64  
 2   Income            2216 non-null   float64
 3   Dt_Customer       2240 non-null   object 
 4   MntWines          2240 non-null   int64  
 5   MntFruits         2240 non-null   int64  
 6   MntMeatProducts   2240 non-null   int64  
 7   MntFishProducts   2240 non-null   int64  
 8   MntSweetProducts  2240 non-null   int64  
 9   MntGoldProds      2240 non-null   int64  
dtypes: float64(1), int64(8), object(1)
memory usage: 175.1+ KB


From the above output, we can conclude and note that:

There are missing values in income
Dt_Customer that indicates the date a customer joined the database is not parsed as DateTime
There are some categorical features in our data frame; as there are some features in dtype: object). So we will need to encode them into numeric forms later.
First of all, for the missing values, I am simply going to drop the rows that have missing income values.

In [5]:
df1 = df1.dropna()
len(df1)

2216

In the next bit,we will be performing the following steps to engineer some new features:

- Extract the "Age" of a customer by the "Year_Birth" indicating the birth year of the respective person.
- Create another feature "Spent" indicating the total amount spent by the customer in various categories over the span of two years.
- Dropping some of the redundant features

In [6]:
df1["Age"] = 2023-df1["Year_Birth"]
df1["Spent"] = df1["MntWines"]+ df1["MntFruits"]+ df1["MntMeatProducts"]+ df1["MntFishProducts"]+ df1["MntSweetProducts"]+ df1["MntGoldProds"]
df1=df1.rename(columns={"MntWines": "Wines","MntFruits":"Fruits","MntMeatProducts":"Meat","MntFishProducts":"Fish","MntSweetProducts":"Sweets","MntGoldProds":"Gold"})
to_drop = ["ID"]
df1 = df1.drop(to_drop, axis=1)

C:\Users\DELL\AppData\Local\Temp\ipykernel_15864\369905042.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1["Age"] = 2023-df1["Year_Birth"]
C:\Users\DELL\AppData\Local\Temp\ipykernel_15864\369905042.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1["Spent"] = df1["MntWines"]+ df1["MntFruits"]+ df1["MntMeatProducts"]+ df1["MntFishProducts"]+ df1["MntSweetProducts"]+ df1["MntGoldProds"]


In [7]:
df.describe()

ID   Year_Birth         Income      Kidhome     Teenhome  \
count   2240.000000  2240.000000    2216.000000  2240.000000  2240.000000   
mean    5592.159821  1968.805804   52247.251354     0.444196     0.506250   
std     3246.662198    11.984069   25173.076661     0.538398     0.544538   
min        0.000000  1893.000000    1730.000000     0.000000     0.000000   
25%     2828.250000  1959.000000   35303.000000     0.000000     0.000000   
50%     5458.500000  1970.000000   51381.500000     0.000000     0.000000   
75%     8427.750000  1977.000000   68522.000000     1.000000     1.000000   
max    11191.000000  1996.000000  666666.000000     2.000000     2.000000   

           Recency     MntWines    MntFruits  MntMeatProducts  \
count  2240.000000  2240.000000  2240.000000      2240.000000   
mean     49.109375   303.935714    26.302232       166.950000   
std      28.962453   336.597393    39.773434       225.715373   
min       0.000000     0.000000     0.000000         0.000000   
25%      24.000000    23.750000     1.000000        16.000000   
50%      49.000000   173.500000     8.000000        67.000000   
75%      74.000000   504.250000    33.000000       232.000000   
max      99.000000  1493.000000   199.000000      1725.000000   

       MntFishProducts  ...  NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  \
count      2240.000000  ...        2240.000000   2240.000000   2240.000000   
mean         37.525446  ...           5.316518      0.072768      0.074554   
std          54.628979  ...           2.426645      0.259813      0.262728   
min           0.000000  ...           0.000000      0.000000      0.000000   
25%           3.000000  ...           3.000000      0.000000      0.000000   
50%          12.000000  ...           6.000000      0.000000      0.000000   
75%          50.000000  ...           7.000000      0.000000      0.000000   
max         259.000000  ...          20.000000      1.000000      1.000000   

       AcceptedCmp5  AcceptedCmp1  AcceptedCmp2     Complain  Z_CostContact  \
count   2240.000000   2240.000000   2240.000000  2240.000000         2240.0   
mean       0.072768      0.064286      0.013393     0.009375            3.0   
std        0.259813      0.245316      0.114976     0.096391            0.0   
min        0.000000      0.000000      0.000000     0.000000            3.0   
25%        0.000000      0.000000      0.000000     0.000000            3.0   
50%        0.000000      0.000000      0.000000     0.000000            3.0   
75%        0.000000      0.000000      0.000000     0.000000            3.0   
max        1.000000      1.000000      1.000000     1.000000            3.0   

       Z_Revenue     Response  
count     2240.0  2240.000000  
mean        11.0     0.149107  
std          0.0     0.356274  
min         11.0     0.000000  
25%         11.0     0.000000  
50%         11.0     0.000000  
75%         11.0     0.000000  
max         11.0     1.000000  

[8 rows x 26 columns]

We filter the year from Dt_Customer to get the data for a line chart

In [8]:
df1['Dt_Customer'] = pd.to_datetime(df1['Dt_Customer'], format='%d-%m-%Y')
df1['Year'] = df1['Dt_Customer'].dt.year
df1['Year'] = df1['Year'].astype(str)
query5 = "SELECT * FROM df1"
dfl = pysqldf(query5)

df1

Year_Birth   Income Dt_Customer  Wines  Fruits  Meat  Fish  Sweets  \
0           1957  58138.0  2012-09-04    635      88   546   172      88   
1           1954  46344.0  2014-03-08     11       1     6     2       1   
2           1965  71613.0  2013-08-21    426      49   127   111      21   
3           1984  26646.0  2014-02-10     11       4    20    10       3   
4           1981  58293.0  2014-01-19    173      43   118    46      27   
...          ...      ...         ...    ...     ...   ...   ...     ...   
2235        1967  61223.0  2013-06-13    709      43   182    42     118   
2236        1946  64014.0  2014-06-10    406       0    30     0       0   
2237        1981  56981.0  2014-01-25    908      48   217    32      12   
2238        1956  69245.0  2014-01-24    428      30   214    80      30   
2239        1954  52869.0  2012-10-15     84       3    61     2       1   

      Gold  Age  Spent  Year  
0       88   66   1617  2012  
1        6   69     27  2014  
2       42   58    776  2013  
3        5   39     53  2014  
4       15   42    422  2014  
...    ...  ...    ...   ...  
2235   247   56   1341  2013  
2236     8   77    444  2014  
2237    24   42   1241  2014  
2238    61   67    843  2014  
2239    21   69    172  2012  

[2216 rows x 12 columns]

In [11]:
query2 = '''
SELECT
    Age,
    COUNT(*) AS count,
    SUM(Wines) AS tl_wines,
    SUM(Fruits) AS tl_frts,
    SUM(Meat) AS tl_meat,
    SUM(Fish) AS tl_fish,
    SUM(Sweets) AS tl_sweet,
    SUM(Gold) AS tl_gold
FROM 
    (SELECT 
        CASE
            WHEN Age BETWEEN 18 AND 30 THEN '18-30'
            WHEN Age BETWEEN 30 AND 49 THEN '30-49'
            WHEN Age BETWEEN 50 AND 64 THEN '50-64'
            WHEN Age BETWEEN 65 AND 75 THEN '65-75'
            WHEN Age BETWEEN 75 AND 89 THEN '75-89'
            WHEN Age BETWEEN 90 AND 100 THEN '90-100'
            ELSE 'Other'
        END AS Age,
            Wines, 
            Fruits, 
            Meat, 
            Fish, 
            Sweets, 
            Gold
    FROM 
        df1) AS subquery
GROUP BY 
    Age;
'''

dfcl1 = sqldf(query2, locals())

dfcl1

Age  count  tl_wines  tl_frts  tl_meat  tl_fish  tl_sweet  tl_gold
0  18-30     15      5357      649     5127     1406       691     1042
1  30-49    826    202450    21170   129077    29659     21945    32879
2  50-64    863    272449    22574   135849    29863     21941    37472
3  65-75    454    167373    11560    83197    18908     13064    23312
4  75-89     55     27676     2302    16238     3451      2187     2471
5  Other      3       778      150      575      118        68      251

In [12]:
df_liner=pysqldf("SELECT Age, Income, Wines, Fruits, Meat, Fish, Sweets, Gold, Spent\
             FROM df1\
             GROUP BY Age\
             ORDER BY Age ASC")
df_liner

Age   Income  Wines  Fruits  Meat  Fish  Sweets  Gold  Spent
0    27   7500.0      3      18    14    15      22    50    122
1    28  34824.0      4       2    11     2       0     4     23
2    29  80134.0    966      26   282    52      26    26   1378
3    30  74293.0    375     152   335    93      91    81   1127
4    31  92859.0    962      61   921    52      61    20   2077
5    32  68126.0   1332      17   311    23      51    86   1820
6    33  73687.0    559     153   503   218      13   181   1627
7    34  38360.0     36       2    42    20      21    10    131
8    35  27938.0     13       0     8     2       1     5     29
9    36  17323.0      3      14    17     6       1     5     46
10   37  34320.0      4       1     9     3       0     0     17
11   38  33454.0     76      10    56     3       1    23    169
12   39  26646.0     11       4    20    10       3     5     53
13   40  30096.0      5       3    11    12       5     9     45
14   41  37040.0     86       2    73    69      38    48    316
15   42  58293.0    173      43   118    46      27    15    422
16   43  41850.0     53       5    19     2      13     4     96
17   44   2447.0      1       1  1725     1       1     1   1730
18   45  66373.0    328       9   124    12      24   109    606
19   46  35790.0     12       6    20    30       1     3     72
20   47   7500.0      6      16    11    11       1    16     61
21   48   7500.0      3       1    10     3       2    12     31
22   49  30351.0     14       0    24     3       3     2     46
23   50  42429.0     55       0     6     2       0     4     67
24   51  75251.0    721     111   925    97      18    18   1890
25   52  55635.0    235      65   164    50      49    27    590
26   53  40548.0    110       0     5     2       0     3    120
27   54  18589.0      6       4    25    15      12    13     75
28   55  41728.0     13       6    15     3       5    13     55
29   56  62513.0    520      42    98     0      42    14    716
30   57  70287.0    295      35   482   121     120    40   1093
31   58  71613.0    426      49   127   111      21    42    776
32   59  79143.0    650      37   780    27     167    32   1693
33   60  38620.0    112      17    44    34      22    89    318
34   61  50785.0     64       1    21     0       1    27    114
35   62  28249.0      1       9     7     2      14    10     43
36   63  29440.0     17       8    14    10       4     3     56
37   64  63033.0    194      61   480   225     112    30   1102
38   65  69096.0    247      49   159     7      82   109    653
39   66  58138.0    635      88   546   172      88    88   1617
40   67  77376.0    492      19   110    16      12    12    661
41   68  82384.0    984      51   432   180     120   190   1957
42   69  46344.0     11       1     6     2       1     6     27
43   70  56129.0    320      48   133    39      72    48    660
44   71  59354.0    233       2    53     3       5    14    310
45   72  40689.0    270       3    27    39       6    99    444
46   73   5648.0     28       0     6     1       1    13     49
47   74  76995.0   1012      80   498     0      16   176   1782
48   75  70666.0    398      40   367   119     122    30   1076
49   76  77457.0    675      11   400    15      45    11   1157
50   77  82800.0   1006      22   115    59      68    45   1315
51   78  65846.0    562      81   276    80      40    81   1120
52   79  55614.0    504      13   131     8       0    26    682
53   80  48948.0    437       8   206   160      49    42    902
54   82  93027.0   1285      42   716    55       0    21   2119
55   83  51141.0    144       0     7     0       0     6    157
56  123  36640.0     15       6     8     7       4    25     65
57  124  83532.0    755     144   562   104      64   224   1853
58  130  60182.0      8       0     5     7       0     2     22

In [25]:
from sklearn.linear_model import LinearRegression
est = LinearRegression(fit_intercept=True)
x=df_liner[['Wines','Fruits','Meat', 'Fish', 'Sweets', 'Gold']]
y=df_liner[['Income']]
est.fit(x,y)
print(est.coef_)
print(est.intercept_)
print("R^2 is ",est.score(pd.DataFrame(X_test),pd.DataFrame(y_test)))

[[ 46.93021094 176.38438793 -13.06163772  -0.28823358 133.77590578
  -65.05377997]]
[32295.44844568]
R^2 is  0.7618912813076114


The R^2 is an evaluation metric that measures the goodness of fit of the model to the data. It represents the proportion of the variance in the target variable that can be explained by the model. R^2 values range from 0 to 1, and a value closer to 1 indicates a better fit of the model to the data. And our R^2 is 0.7618912813076114 so we can see that the model was fit to the data.

In [14]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)
est = LinearRegression(fit_intercept=True)
est.fit(pd.DataFrame(X_train),pd.DataFrame(y_train))

LinearRegression()

In [15]:
import statsmodels.api as sm
X_train2 = sm.add_constant(X_train) # This adds an intercept
model = sm.OLS(y_train,X_train2)
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Income   R-squared:                       0.734
Model:                            OLS   Adj. R-squared:                  0.694
Method:                 Least Squares   F-statistic:                     18.36
Date:                Sun, 08 Oct 2023   Prob (F-statistic):           4.18e-10
Time:                        17:38:12   Log-Likelihood:                -507.96
No. Observations:                  47   AIC:                             1030.
Df Residuals:                      40   BIC:                             1043.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3.227e+04   2864.627     11.267      0.000    2.65e+04    3.81e+04
Wines         43.7058      6.814      6.414      0.000      29.935      57.477
Fruits       172.3178     71.442      2.412      0.021      27.928     316.708
Meat         -13.7021      6.939     -1.975      0.055     -27.727       0.323
Fish           1.2796     44.220      0.029      0.977     -88.093      90.652
Sweets       142.1628     57.648      2.466      0.018      25.652     258.674
Gold         -50.8173     48.952     -1.038      0.305    -149.753      48.118
==============================================================================
Omnibus:                        2.627   Durbin-Watson:                   1.896
Prob(Omnibus):                  0.269   Jarque-Bera (JB):                1.705
Skew:                          -0.428   Prob(JB):                        0.426
Kurtosis:                       3.371   Cond. No.                         884.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Coefficients: The coefficients of the model indicate the relationship between the independent variables and the target variable. In linear regression, these coefficients represent the change in the target variable for a one-unit change in the corresponding independent variable. A higher magnitude coefficient suggests a stronger impact on the target variable.

Intercept: The intercept represents the estimated target variable value when all independent variables are zero. It provides a baseline reference point for the model's predictions.

The p-value is 4.18e-10, which is a very small value (smaller than 0.05 or 0.01), indicating that the F-statistic is highly statistically significant. This suggests that the overall regression model has a significant relationship with the dependent variable, and at least one independent variable in the model is contributing significantly to explaining the variation in the dependent variable. The regression model is statistically significant and provides a good fit for the data.

R-squared (R2) Score: R2 score measures the proportion of the variance in the target variable that can be explained by the independent variables. It ranges from 0 to 1, with higher values indicating a better fit. R2 score of 1 means the model explains all the variance in the target variable, while a score of 0 means the model fails to explain any variance.

In [18]:
dfclean1=pysqldf("SELECT COUNT(Age), Age\
             FROM df1\
             GROUP BY Age\
             ORDER BY Age ASC")
dfclean1.head()

COUNT(Age)  Age
0           2   27
1           5   28
2           3   29
3           5   30
4          13   31